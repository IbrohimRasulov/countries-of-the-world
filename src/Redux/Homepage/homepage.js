const GET_COUNTRIES_API = 'CountriesOfTheWorld/countries/GET_COUNTRIES_API';
const COUNTRIES_API = 'https://restcountries.com/v2/all';

const initialState = [];

export const getCountries = (payload) => ({
  type: GET_COUNTRIES_API,
  payload,
});

export const getCountriesFromAPI = () => async (dispatch) => {
  await fetch(COUNTRIES_API)
    .then((response) => response.json())
    .then((data) => {
      const arrData = data.map((country) => ({
        name: country.name,
        flag: country.flags.png,
        region: country.region,
      }));
      if (arrData) {
        dispatch(getCountries(arrData));
      }
    });
};

const homepageReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_COUNTRIES_API:
      return [...action.payload];
    default:
      return state;
  }
};

export default homepageReducer;
