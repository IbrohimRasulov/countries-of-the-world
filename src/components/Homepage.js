import world_map from '../assets/World-map.jpg';
import { useSelector, useDispatch } from 'react-redux';
import { useEffect, useState } from 'react';
import { getCountriesFromAPI } from '../Redux/Homepage/homepage';
import { Link } from 'react-router-dom';
import Header from './Header';

const Homepage = () => {
  const countriesList = useSelector((state) => state.homepageReducer)
  const dispatch = useDispatch();
  const [search, setSearch] = useState('');

  useEffect(() => {
    dispatch(getCountriesFromAPI());
  }, [dispatch]);

  const filteredData = countriesList.filter((country) => {
    const data = Object.keys(country).some((key) => {
      const res = country[key].toLowerCase().includes(search.toLowerCase());
      return res;
    });
    return data;
  });

  return (
    <div className="homepage">
      <Header className="App-header" search="true" />
      <input type="text" placeholder="Search..." onChange={(e) => setSearch(e.target.value)} />
      <div className="overview">
        <img src={world_map} alt="World map" />
        {filteredData.length !== 0 && <h1>Countries Of The World</h1>}
      </div>
      <ul className="list">
        {countriesList.length === 0 && <h1 className="loading">Loading...</h1>}
        {(filteredData.length === 0 && search.length !== 0) && <h1 className="loading">No match :(</h1>}
        {filteredData && filteredData.map((country) => {
          return (
            <li key={country.name}>
              <Link to={country.name}>
                <img src={country.flag} alt="Flag" />
                <p>{country.name}</p>
              </Link>
            </li>
          );
        })}
      </ul>
    </div>
  );
};

export default Homepage;
