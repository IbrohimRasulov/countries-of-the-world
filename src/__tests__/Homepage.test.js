import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../Redux/configureStore';
import Homepage from '../components/Homepage';

it('testing rendering Homepage component', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <Homepage />
      </Router>
    </Provider>,
  ).toJSON;
  expect(tree).toMatchSnapshot();
});
