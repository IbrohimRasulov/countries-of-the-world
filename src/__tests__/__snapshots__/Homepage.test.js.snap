// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing rendering Homepage component 1`] = `[Function]`;

exports[`testing rendering Profile component 1`] = `[Function]`;
